### PR TITLE
ENH: Implement RotationSpline

### PR DESCRIPTION
**Pull request by nmayorov**
I decided to finish up #9024

I used a self-written reference for the implementation  
[attitude_interpolation.pdf](https://github.com/scipy/scipy/files/2932755/attitude_interpolation.pdf)

Which is more clear than what can be found in the original reference.

@adbugger --- would be very cool if you could review/participate in this PR.

### CODE REVIEW

**Comment by rgommers** [comment]
> I used a self-written reference for the implementation
[attitude_interpolation.pdf](https://github.com/scipy/scipy/files/2873382/attitude_interpolation.pdf) Which is more clear than what can be found in the original reference.

Nice. I was going to suggest to put that somewhere public/permanent, but then I saw that the link is to somewhere within this repo (`https://github.com/scipy/scipy/files/2873382/attitude_interpolation.pdf`). How did you get that file there?

**Comment by nmayorov** [comment]
@rgommers 

Well, I just dragged-and-drooped the file while editing the text 😃I agree that it's helpful to make these notes publicly available, I'm going to clean them up a little bit.

**Comment by nmayorov** [comment]
@rgommers, could you suggest how we can proceed with this PR? I'm happy with the code overall.

**Comment by rgommers** [comment]
> I decided to finish up #9024

You rewrote from scratch right, rather than continuing that PR? (asking because of author attribution)

> @rgommers, could you suggest how we can proceed with this PR? I'm happy with the code overall.

It looks pretty good at first sight. I was hoping @adbugger was still interested to review/participate. Aditya, please let us know - also if you're not interested or simply have no time at the moment, would be good to hear from you.

Otherwise I'll try and make time to review soon. Marking it for the next release for now.

**Comment by nmayorov** [comment]
> You rewrote from scratch right, rather than continuing that PR? (asking because of author attribution)

I have mostly rewritten it, but used some structure from @adbugger version. 

I'm going to update the file above (very minor cleanup) and reference it in the docstring.

**Comment by adbugger** [comment]
Apologies for the delay, I'm working on a few other research projects at the moment so saw the notifications now.

Looks good to me upon review. Only the structure of the `RotationSpline` class is loosely based on my original code. The rest is expanded and re-written in a much clearer way, and using `scipy.linalg.solve_banded` to do some of the heavy lifting is great. I should've looked into other helpful scipy functions sooner.

Thanks @nmayorov for finishing this. Once again, I apologize for the delayed replies.

**Comment by rgommers** [comment]
@adbugger no worries, good to hear from you. come look us up when you have time!

File: scipy/spatial/transform/_rotation_spline.py [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+            2 * A_inv[1:-1] / dt[1:-1, None, None],
+            2 * A[1:-1] / dt[1:-1, None, None],
+            4 * (1 / dt[:-1] + 1 / dt[1:]))
+
+        b0 = 6 * (rotvecs[:-1] * dt[:-1, None] ** -2 +
+                  rotvecs[1:] * dt[1:, None] ** -2)
+        b0[0] -= 2 / dt[0] * A_inv[0].dot(angular_rate_first)
+        b0[-1] -= 2 / dt[-1] * A[-1].dot(angular_rates[-1])
+
+        for iteration in range(self.MAX_ITER):
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: partial coverage hit on this line; perhaps safe to ignore--probably means this loop tends to break

File: scipy/spatial/transform/_rotation_spline.py [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+        rotvecs_dot = _matrix_vector_product_of_stacks(A, angular_rates)
+        angular_rates = np.vstack((angular_rate_first, angular_rates[:-1]))
+
+        return angular_rates, rotvecs_dot
+
+    def __init__(self, times, rotations):
+        from scipy.interpolate import PPoly
+
+        if len(rotations) == 1:
+            raise ValueError("`rotations` must contain at least 2 rotations.")
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: coverage miss

File: scipy/spatial/transform/_rotation_spline.py [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+
+        times = np.asarray(times, dtype=float)
+        if times.ndim != 1:
+            raise ValueError("`times` must be 1-dimensional.")
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: coverage miss

File: scipy/spatial/transform/_rotation_spline.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+
+        if len(times) != len(rotations):
+            raise ValueError("Expected number of rotations to be equal to "
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: coverage miss

File: scipy/spatial/transform/_rotation_spline.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+                             "number of timestamps given, got {} rotations "
+                             "and {} timestamps."
+                             .format(len(rotations), len(times)))
+
+        dt = np.diff(times)
+        if np.any(dt <= 0):
+            raise ValueError("Values in `times` must be in a strictly "
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: coverage miss

File: scipy/spatial/transform/_rotation_spline.py [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+                * 0 (default) : return Rotation
+                * 1 : return the angular rate in rad/sec
+                * 2 : return the angular acceleration in rad/sec/sec
+
+        Returns
+        -------
+        Interpolated Rotation, angular rate or acceleration.
+        """
+        if order not in [0, 1, 2]:
+            raise ValueError("`order` must be 0, 1 or 2.")
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: coverage miss

File: scipy/spatial/transform/_rotation_spline.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+
+        times = np.asarray(times, dtype=float)
+        if times.ndim > 1:
+            raise ValueError("`times` must be at most 1-dimensional.")
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: coverage miss

File: scipy/spatial/transform/_rotation_spline.py [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+            result = self.rotations[index] * Rotation.from_rotvec(rotvecs)
+        elif order == 1:
+            rotvecs_dot = self.interpolator(times, 1)
+            result = _compute_angular_rate(rotvecs, rotvecs_dot)
+        elif order == 2:
+            rotvecs_dot = self.interpolator(times, 1)
+            rotvecs_dot_dot = self.interpolator(times, 2)
+            result = _compute_angular_acceleration(rotvecs, rotvecs_dot,
+                                                   rotvecs_dot_dot)
+        else:
```
Comment (tylerjereddy) [inline scipy/spatial/transform/_rotation_spline.py]: control flow never reaches "else" in current tests

**Review by tylerjereddy — COMMENTED** [review]
Added a few initial / superficial comments from running coverage locally.

Not related to this PR, but with codecov "dead for SciPy" we may want to look at just storing coverage run results in an html artifact on Azure or something like that. They have docs about publishing cov results.

**Comment by tylerjereddy** [comment]
I'll bump the milestone. I really should review this in more detail myself beyond the coverage comments above since it is in spatial, but for a 600 line diff we may also not want to rush it before Thursday.

File: scipy/spatial/transform/_rotation_spline.py (hunk 1 above)
Comment (nmayorov) [reply scipy/spatial/transform/_rotation_spline.py]: You are right about that.

File: scipy/spatial/transform/_rotation_spline.py (hunk 8 above)
Comment (nmayorov) [reply scipy/spatial/transform/_rotation_spline.py]: "else" contains `assert False`, because `order` must be 0, 1 or 2 by that point.

**Comment by nmayorov** [comment]
@tylerjereddy 

I improved the coverage to 100%.

> I'll bump the milestone. I really should review this in more detail myself beyond the coverage comments above since it is in spatial, but for a 600 line diff we may also not want to rush it before Thursday.

It's a bit unfortunate, but OK, if you think it is necessary this way.

**Comment by rgommers** [comment]
It's still Thursday. I'm comfortable merging this as is for 1.3.0, it's in good shape and @adbugger who worked on this before has also reviewed.

**Comment by rgommers** [comment]
@tylerjereddy if you're not in the middle of branching yet, can you comment or press the green button?

**Comment by tylerjereddy** [comment]
alright done; I'm polishing release notes & mailmap at the moment

**Comment by tylerjereddy** [comment]
Thanks @nmayorov & reviewers

**Pull request merged**